drivers/power_domain/power_domain_device_onoff.c

```diff
@@ -0,0 +1,123 @@
+	/* In case device is active, first suspend it before turning it off */
+	if ((visitor_context->action == PM_DEVICE_ACTION_TURN_OFF) &&
+		(dev->pm->state == PM_DEVICE_STATE_ACTIVE)) {
+		ret = pm_device_action_run(dev, PM_DEVICE_ACTION_SUSPEND);
+		/* Check the type of failure */
+		if (ret != 0) {
+			if ((ret != -ENOSYS) || (ret != -ENOTSUP) || (ret != -EALREADY)) {
+				return ret;
+			}
+		}
+	}
+	(void)pm_device_action_run(dev, visitor_context->action);
+	return 0;
+}
+
+static int pd_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct pd_deviceonoff_config *config = dev->config;
+	uint8_t i = 0;
+	int ret = 0;
+	enum pm_state state = pm_state_next_get(0)->state;
+	struct pd_visitor_context context = {.domain = dev};
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		while (config->onoff_power_states[i] != POWER_DOMAIN_DEVICE_ONOFF_STATE_COUNT) {
+			if (state == config->onoff_power_states[i]) {
+				/* Notify devices on the domain they are now powered */
+				context.action = PM_DEVICE_ACTION_TURN_ON;
+				ret = device_supported_foreach(dev, pd_domain_visitor, &context);
+				/* No need to go through the rest of the array of states */
+				break;
+			}
+			i++;
+		}
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		while (config->onoff_power_states[i] != POWER_DOMAIN_DEVICE_ONOFF_STATE_COUNT) {
+			if (state == config->onoff_power_states[i]) {
+				/* Notify devices on the domain that power is going down */
+				context.action = PM_DEVICE_ACTION_TURN_OFF;
+				ret = device_supported_foreach(dev, pd_domain_visitor, &context);
+				/* No need to go through the rest of the array of states */
+				break;
+			}
+			i++;
+		}
+		break;
+	case PM_DEVICE_ACTION_TURN_ON:
+		break;
+	case PM_DEVICE_ACTION_TURN_OFF:
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return ret;
+}
+
+#define DT_DRV_COMPAT power_domain_deviceonoff
+
+#define PM_STATE_FROM_DT(i, node_id, prop_name)						\
+	COND_CODE_1(DT_NODE_HAS_STATUS(DT_PHANDLE_BY_IDX(node_id, prop_name, i), okay),	\
+		(PM_STATE_DT_INIT(DT_PHANDLE_BY_IDX(node_id, prop_name, i)),), ())
+
+#define PM_STATE_FROM_DT_2(i, node_id, prop_name)					\
+		PM_STATE_DT_INIT(DT_PHANDLE_BY_IDX(node_id, prop_name, i))
+
+#define POWER_DOMAIN_DEVICE_ONOFF_STATES(inst, node_id)					\
+	uint8_t onoff_states_##inst[] = {						\
+		LISTIFY(DT_PROP_LEN_OR(node_id, onoff_power_states, 0),			\
+			PM_STATE_FROM_DT, (), node_id, onoff_power_states)		\
+		POWER_DOMAIN_DEVICE_ONOFF_STATE_COUNT					\
+	};
+
+#define POWER_DOMAIN_DEVICE(id)								\
+	POWER_DOMAIN_DEVICE_ONOFF_STATES(id, DT_DRV_INST(id))				\
+											\
+	static const struct pd_deviceonoff_config pd_deviceonoff_##id##_cfg = {		\
+		.onoff_power_states = onoff_states_##id,				\
+	};										\
+	PM_DEVICE_DT_INST_DEFINE(id, pd_pm_action);					\
+	DEVICE_DT_INST_DEFINE(id, NULL, PM_DEVICE_DT_INST_GET(id),			\
+			      NULL, &pd_deviceonoff_##id##_cfg, PRE_KERNEL_1,		\
+			      CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
+
```

Comment:
@mmahadevan108 I think this domain is re-implementing a lot of things from this pr https://github.com/zephyrproject-rtos/zephyr/pull/60939

This logic to check which power state should trigger device pm should be done in the system pm layer imho.

Wondering now if we should have add a way to the system pm triggers POWER_ON / POWER_OFF depending on the sleep state the soc goes (instead of just suspend / resume).